### ISSUE TITLE
💅 `noFloatingPromise` does not catch missing promises called through generically typed callback parameters

### ISSUE THREAD

**Issue by glassmancody-lcm** [biomejs/biome#11605]
### Environment information

<details>

```bash
CLI:
  Version:                      2.4.6
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.14.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.11.0

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  false
  Enabled rules:
    complexity/noAdjacentSpacesInRegex
    complexity/noExtraBooleanCast
    complexity/noUselessCatch
    complexity/noUselessEscapeInRegex
    complexity/noUselessTypeConstraint
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noSwitchDeclarations
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noFloatingPromises
    style/noEnum
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useConsistentMemberAccessibility
    style/useFilenamingConvention
    style/useNamingConvention
    style/useNumericSeparators
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCompareNegZero
    suspicious/noConstantBinaryExpressions
    suspicious/noDebugger
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noImportAssign
    suspicious/noIrregularWhitespace
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noShadowRestrictedNames
    suspicious/noSparseArray
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useGetterReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

</details>

### Rule name

noFloatingPromises

### Playground link

[Playground](https://biomejs.dev/playground/?lintRules=noFloatingPromises&tab=formatter&pane=Diagnostics#code=aQBuAHQAZQByAGYAYQBjAGUAIABTAGUAcgB2AGkAYwBlACAAewAKACAAIAAgACAAZwBvACgAKQA6ACAAUAByAG8AbQBpAHMAZQA8AHYAbwBpAGQAPgA7AAoAfQAKAGkAbgB0AGUAcgBmAGEAYwBlACAAQwBvAG4AdABlAHgAdAAgAHsACgAgACAAIAAgAHMAZQByAHYAaQBjAGUAOgAgAFMAZQByAHYAaQBjAGUAOwAKAH0ACgAKAGQAZQBjAGwAYQByAGUAIABmAHUAbgBjAHQAaQBvAG4AIAB0AGUAcwB0ADwAVAA%2BACgACgAgACAAIAAgAGMAYQBsAGwAYgBhAGMAawA6ACAAKABhAHIAZwBzADoAIABUACkAIAA9AD4AIABQAHIAbwBtAGkAcwBlADwAdgBvAGkAZAA%2BACwACgApADoAIAB2AG8AaQBkADsACgAKAHQAZQBzAHQAPABDAG8AbgB0AGUAeAB0AD4AKABhAHMAeQBuAGMAIAAoAGEAcgBnAHMAKQAgAD0APgAgAHsACgAgACAAIAAgAC8ALwAgAFQAaABpAHMAIABjAGEAdABjAGgAZQBzACAAbQBpAHMAcwBpAG4AZwAgAGEAdwBhAGkAdAAgAGkAZgAgAHUAbgBjAG8AbQBtAGUAbgB0AGUAZAAKACAAIAAgACAALwAvACAAYwBvAG4AcwB0ACAAcwBlAHIAdgBpAGMAZQA6ACAAUwBlAHIAdgBpAGMAZQAgAD0AIABhAHIAZwBzAC4AcwBlAHIAdgBpAGMAZQA7AAoAIAAgACAAIABjAG8AbgBzAHQAIABzAGUAcgB2AGkAYwBlACAAPQAgAGEAcgBnAHMALgBzAGUAcgB2AGkAYwBlADsACgAgACAAIAAgAC8ALwAgAEUAeABwAGUAYwB0AGUAZAAgAHQAbwAgAHQAaAByAG8AdwAgAGUAcgByAG8AcgAgAGgAZQByAGUALAAgAHQAaABlAHIAZQAgAGkAcwAgAGEAIABtAGkAcwBzAGkAbgBnACAAYQB3AGEAaQB0AAoAIAAgACAAIABzAGUAcgB2AGkAYwBlAC4AZwBvACgAKQA7AAoAfQApAA%3D%3D
)
### Expected result

After enabling the `noFloatingPromises` linting rule I noticed a lot of missed cases when using it for our Playwright tests. After a little investigation I was able to reproduce the issue with a standalone code snippet:

Here, the missing await is caught by the linting rule as expected.

```typescript
interface Service {
    go(): Promise<void>;
}
interface Context {
    service: Service;
}

declare function test<T>(
    callback: (args: T) => Promise<void>,
): void;

test<Context>(async (args) => {
    const service: Service = args.service;
    service.go();
})
```

Here, the the missing await is _not_ caught by the linting rule. I would expect biome to error out here and catch the missing `await`.

```typescript
interface Service {
    go(): Promise<void>;
}
interface Context {
    service: Service;
}

declare function test<T>(
    callback: (args: T) => Promise<void>,
): void;

test<Context>(async (args) => {
    const service = args.service;
    service.go();
})
```

It seems to me there is some type inference that's being lost, and it might not be necessarily tied to the linting rule in particular. 

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
A slightly reduced reproduction:

```ts
interface Service {
    go(): Promise<void>;
}

interface Context {
    service: Service;
}

// note the missing generic <T>
declare function test(
    callback: (args: Context) => Promise<void>,
): void;

test(async (args) => {
    const service = args.service;
    service.go();
});
```

so the bug doesn't fully have to do with generic parameters
